[13.0][MIG] delivery_gls_asm: Migration to 13.0

victoralmau opened the pull request:
Migración a 13.0

Bloqueado por:

- [x] `delivery_package_number` https://github.com/OCA/delivery-carrier/pull/339

Por favor @chienandalu y @joao-p-marques ¿podéis revisarlo?

@Tecnativa TT26916

chienandalu committed:
[ADD] gls_asm_delivery: New module

oca-travis committed:
[UPD] Update delivery_gls_asm.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: l10n-spain-12.0/l10n-spain-12.0-delivery_gls_asm
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-12-0/l10n-spain-12-0-delivery_gls_asm/

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: l10n-spain-12.0/l10n-spain-12.0-delivery_gls_asm
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-12-0/l10n-spain-12-0-delivery_gls_asm/

chienandalu committed:
Translated using Weblate (Spanish)

Currently translated at 68.4% (93 of 136 strings)

Translation: l10n-spain-12.0/l10n-spain-12.0-delivery_gls_asm
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-12-0/l10n-spain-12-0-delivery_gls_asm/es/

chienandalu committed:
[IMP] delivery_gls and delivery_seur: add informative exceptions

Non implemented methods should raise informative exceptions if used

oca-travis committed:
[UPD] Update delivery_gls_asm.pot

OCA-git-bot committed:
delivery_gls_asm 12.0.1.0.1

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: l10n-spain-12.0/l10n-spain-12.0-delivery_gls_asm
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-12-0/l10n-spain-12-0-delivery_gls_asm/

chienandalu committed:
[FIX] delivery_gls_asm: manifest

OCA-git-bot committed:
delivery_gls_asm 12.0.1.0.2

chienandalu committed:
[FIX] delivery_gls_manifest: manifest destination name

The `cliente` key corresponds to the company not to the customer, wich
is `nombre_dst`

OCA-git-bot committed:
delivery_gls_asm 12.0.1.0.3

pedrobaeza committed:
[FIX] delivery_gls_asm: Proper signature for gls_asm_rate_shipment

Alhough it will raise another exception, at least it will show the
proper message and not a traceback.

victoralmau committed:
[IMP] delivery_gls_asm: black, isort, prettier

victoralmau committed:
[MIG] delivery_gls_asm: Migration to 13.0

joao-p-marques commented on delivery_gls_asm/__manifest__.py:
```diff
@@ -3,15 +3,15 @@
-    "depends": ["delivery_package_number", "delivery_state",],
-    "external_dependencies": {"python": ["suds"],},
+    "depends": ["delivery_package_number", "delivery_state"],
+    "external_dependencies": {"python": ["suds"]},
```
Esto debería estar con el nombre que se usa en Pypi. Por lo que veo en `requirements.txt`, sería `suds-py3`, no?
Si estas librerías no son la misma, por lo menos deberían estar unificadas las dependencias en el manifiesto y las de `requirements.txt`

joao-p-marques reviewed (changes requested):
Revisión de código y me parece bien :+1: . Solo 1 comentario y revisa Travis porfa.

chienandalu reviewed (commented):
I'm testing it with the GLS test user but I can't make it work. Did you achieve to record shippings?

victoralmau commented:
I use tests only but i don't know exactly if it's need to change something to create shippings in GLS, maybe  https://github.com/OCA/l10n-spain/blob/12.0/delivery_gls_asm/data/delivery_asm_data.xml it's necessary to updated?.

chienandalu commented:
> I use tests only but i don't know exactly if it's need to change something to create shippings in GLS, maybe https://github.com/OCA/l10n-spain/blob/12.0/delivery_gls_asm/data/delivery_asm_data.xml it's necessary to updated?.

El problema es que los tests hacen precisamente un mockup de la clase para evitar estar conectando a GLS. Sí, ese es el usuario demo y en principio funcionaría tal cual. Prueba en tu local a hacer una expedición. Mi sospecha va a por la dependencia de suds que has puesto, ya que la API de ASM es bastante quisquillosa y no se lleva bien con cualquier librería SOAP.

victoralmau commented:
He revisado en local a realizar una expedición y funciona correctamente por lo que ese cambio de dependencia no afecta al funcionamiento.

pedrobaeza commented:
Si existe un usuario demo que se puede probar, por qué no hacerlo en los propios tests?

chienandalu commented:
¿Queremos que los tests de integración dependan de la disponibilidad de servicio de GLS?

victoralmau commented:
Actualmente estaba cambiando los tests para que funcione contra la api sin mocks (en local funcionan los tests correctamente pero en travis no). 
Si os parece, espero a que decidáis si se dejan los tests con mocks (como está actualmente en 12.0) o se cambian.

pedrobaeza commented:
Esto ya sabeis que siempre hay debate, pero yo soy de la opinión que si falla la API, esto debe fallar, porque al final tus usuarios no van a poder utilizar el módulo si falla la API. ¿De qué sirve hacer mock de algo para engañar? Y aquí sí que es más claro que con Geonames, que solo se utiliza en la importación inicial.

chienandalu committed:
[FIX] delivery_gls_asm: escaping shipping vals for xml

Ensure that the values sent to the xml soap request are properly
escaped. A partner name like "Johnson & Johnson" would fail otherwise.

chienandalu commented:
> Esto ya sabeis que siempre hay debate, pero yo soy de la opinión que si falla la API, esto debe fallar, porque al final tus usuarios no van a poder utilizar el módulo si falla la API

También bloqueariamos el repositorio completo. Algo bastante molesto cuando se trata de problemas temporales, que suele ser la tónica habitual en estas APIs estables.

Ciertamente es interesante testear en real (algo que por cierto no es posible con el módulo de SEUR) y de hecho en este caso nos va a dar una pista de porqué no pasa en el repo.

Ojalá hubiese una posición intermedia de tests que solo se lanzan en modo PR :sweat_smile:

pedrobaeza commented:
Pero yo considero que también debe pasar en cualquier momento, porque así te das cuenta que ocurre. Opciones en ese caso:

- Deshabilitar el test.
- Esperar a que esté solucionado.

Esto puede bloquear un PR no relacionado, pero para eso se puede escoger la opción 1 si corriera prisa.

victoralmau commented:
Creo que ya estaría listo para poder revisar ahora que ya está aplicado el rebase y si procede mezclaros.

chienandalu commented on delivery_gls_asm/tests/test_delivery_gls_asm.py:
```diff
@@ -37,68 +33,33 @@ def setUpClass(cls):
-        ):
-            self.picking.cancel_shipment()
-        self.assertFalse(self.picking.gls_asm_public_tracking_ref)
+        self.picking.name = "TEST-%s" % random.randint(
+            10000, 99999
+        )  # Prevent API error
```
Creo que esto es más aleatorio :slightly_smiling_face: :
```suggestion
        # GLS API prevents duplicated references so in order to test we need a unique key
        # that doesn't collide with any CI around, as every test really records an expedition
        self.picking.name = "ODOO-TEST-{}".format(time.time())
```

chienandalu commented on delivery_gls_asm/tests/test_delivery_gls_asm.py:
```diff
@@ -37,68 +33,33 @@ def setUpClass(cls):
-            report = wizard.get_manifest()
-            self.assertEqual(report["data"]["deliveries"], mocked_response)
+        with self.assertRaises(UserError):
+            wizard.get_manifest()
```
No nos da muchas pistas de que el test sea correcto, ¿no? ¿Porqué no probamos el método de la clase directamente?

victoralmau replied:
Se ha dividido esto en 2 tests, uno que comprueba para el día actual que se obtenga algún dato (el test previo que ha creado uno ya hace que devuelva datos) y por otro lado el error de sin datos.

chienandalu replied:
Uf, pero es un informe muy pesado para recuperarlo en un test. Ten en cuenta que cada día se generan miles de pruebas de distintos orígenes, por lo que a medida que va pasando el día cuesta más obtener el informe. Casi prefiero comprobar simplemente que no falla la llamada.

victoralmau replied:
Ok, quitamos entonces este test (para un futuro quizás, si se pudiera filtrar la llamada a la API para que devolviera un máximo de x registros no sería tan pesado).

chienandalu commented on delivery_gls_asm/tests/test_delivery_gls_asm.py:
```diff
@@ -37,68 +33,33 @@ def setUpClass(cls):
+        """API work although without data"""
         wizard = self.env["gls.asm.minifest.wizard"].create(
-            {"carrier_id": self.carrier_gls_asm.id, "date_from": "2020-05-31",}
+            {"carrier_id": self.carrier_gls_asm.id, "date_from": "2030-05-31"}
```
Ya puestos, vamos a darnos más margen, ¿no? :smile:

victoralmau replied:
He aumentado esto a 2050, la cuestión de esto es comprobar el error y ampliar la cobertura.

chienandalu commented:
Incluye este cambio reciente también, porfa: https://github.com/OCA/l10n-spain/pull/1632

pedrobaeza commented:
@victoralmau ten en cuenta #1637

chienandalu reviewed (approved):
Ok, vamos con ello :)

pedrobaeza commented:
/ocabot merge nobump

pedrobaeza commented:
/ocabot merge nobump

pedrobaeza commented:
/ocabot merge nobump
